### PR TITLE
23878 user profile federation rebase

### LINKED ISSUE #23878: User profile configuration scoped to user-federation provider
### Typical use-case

- I have read-only LDAP server in Keycloak
- Users coming from such LDAP server should not be able to update their attributes (firstName, lastName or email or any other)
- It will be ideal if fields are shown to users as read-only when user is updating his profile in Account console or by "Update profile" required action form. Same applies for administrators as they also should not be able to update such users (Admin console, Admin REST API).
- At the same time, my realm may contain users coming from the Keycloak DB and/or from other user storage provider (for instance from some different LDAP server, which is writable by Keycloak)
- So I want attributes (firstName, lastname, email etc) to be read-only coming from the read-only LDAP. But writable for other providers

### Other notes and questions

- Can we have user-profile configuration, which are "scoped" to particular user-storage provider? For instance attribute `firstName` will have configuration scoped to user-storage provider `my-read-only-ldap`, which will mark it as "read only". There would be also configuration for `firstName` attribute in realm level. Users NOT coming from read-only LDAP provider will have different configuration of `firstName` marked as writable.  Example (for illustration purposes):
```

{
  "name": "firstName",
  "displayName": "${firstName}",
  "permissions": {
    "view": ["admin", "user"],
    "edit": ["admin", "user"]
  },
  "validators": ...

  // This specify that for users coming from read-only LDAP, attribute is read-only. Admin or users cannot edit it
  "config-overrides": [
    "type": "user-storage-provider",
    "name": "my-read-only-ldap-provider-name",
    "config":  {
      "permissions": {
        "view": ["admin", "user"],
        "edit": []
      },
    }
  ]
}
```

Type can refer to user-storage-provider, group, identity-provider etc. For instance `group` to specify that users from group `xy` are read-only etc...

- Can we somehow trigger "auto-update" of user profile when some user-storage provider is created and/or updated? For instance when someone creates read-only LDAP provider, it can be nice if user-storage provider attributes are automatically updated to reflect this. This can possibly also take LDAP mappers into consideration (EG. When I create LDAP mapper for attribute "birthdate", user-profile might automatically create user-storage provider configuration for the attribute "birthdate". But this is to be discussed how exactly this should work...

- Do we want to support the use-case when users coming from user-storage provider `abc` should not be managed and validated by user profile at all?

### PR DESCRIPTION
**Pull request by mposolda**
closes #23878

### CODE REVIEW

**Comment by mposolda** [comment]
@pedroigor Cound you please check this? There are failing tests and also I did not added any tests for this. I can add them if we agree if the approach is good to go.

Added 1 commit on top of your previous 2 commits in this branch. My commit is here https://github.com/keycloak/keycloak/pull/24282/commits/770fef5306d0891777d1f70125a1435f2efc664b, so you can easily review all the changes. Summary of what I did:

- Removed method `decorateUserProfileMetadata(Object metadata)` from `UserProvider` (together with your `TODO`). Instead made `UserCacheSession` and `UserStorageManager` to implement `UserProfileDecorator`, so it is possible to use directly this interface instead of the method with "object" argument, which looked a bit workaroundy to me. WDYT?

- The `UserProfileDecorator` was not used when administrator did not updated default config of user-profile. Did some small changes in `DeclarativeUserProfileProvider` to handle this.

- Removed your changes from `DefaultAttributes` (your very 1st commit, which I believe is obsolete with the new approach of `UserProfileDecorator`?)

- Introduced new group `user-metadata` into `keycloak-user-profile-configuration.json`. The metadata attributes like `LDAP_ID` are added to this group. See screenshot. WDYT?

Screenshot from 2023-10-25 18-36-00


- In case that LDAP server is read-only and user with his attributes is not editable, it is propagated into user-profile configuration as well and hence also to the UI (See another screenshot)

Screenshot from 2023-10-25 18-43-03


- Last change: Introduced the option that LDAP provider adds user attributes to user profile, which are defined in it as mappers. There is new method `getUserAttributes()` on `LDAPStorageMapper`, which allows that mapper can announce the user attributes, which it supports. So based on all the configured LDAP mappers, it is possible to automatically add attributes to user-profile as long as they are defined here. I am not 100% sure if this is clashes with your changes regarding mappers. So if you prefer, I can remove this change from the PR...

File: services/src/main/resources/org/keycloak/userprofile/config/keycloak-default-user-profile.json
Comment (pedroigor) [inline services/src/main/resources/org/keycloak/userprofile/config/keycloak-default-user-profile.json:59]: I'm not sure what is the best name for this group too. But `User metadata` seems too generic and perhaps `LDAP` or `LDAP Metadata` or `LDAP Internal Attributes` or something else along these lines?

Or are you thinking about adding not only LDAP attributes to this group? Like any other internal-use attribute?

File: federation/ldap/src/main/java/org/keycloak/storage/ldap/LDAPStorageProvider.java
Comment (mposolda) [reply federation/ldap/src/main/java/org/keycloak/storage/ldap/LDAPStorageProvider.java:1014]: Thanks, I will change it

File: services/src/main/resources/org/keycloak/userprofile/config/keycloak-default-user-profile.json
Comment (mposolda) [reply services/src/main/resources/org/keycloak/userprofile/config/keycloak-default-user-profile.json:59]: Yes, I was thinking about this as `User metadata`, so it is not only specific to LDAP, but it is possible to add here any other metadata attributes. I am planning to update this PR to handle also `Kerberos provider` and `SSSD provider` and I expect to use same group `User metadata` for those providers as well. Also administrator can possibly use this group to add some attributes manually though? Maybe the attributes like `userCertificate` we discussed before or some others?

File: federation/ldap/src/main/java/org/keycloak/storage/ldap/LDAPStorageProvider.java
Comment (mposolda) [reply federation/ldap/src/main/java/org/keycloak/storage/ldap/LDAPStorageProvider.java:995]: Ok, I will remove annotation from this. I used the annotation because of https://github.com/keycloak/keycloak/issues/24319, which render the fields as "Combobox" . But it is already fixed according to UI team, so I might just need to completely rebuild UI...

I think that if administrators want to change annotations for attributes, they may need to define the particular attribute in user-profile? With my current changes, in case that field is already defined in user-profile, it is not added by user-storage provider due to https://github.com/mposolda/keycloak/blob/23878-user-profile-federation-rebase/federation/ldap/src/main/java/org/keycloak/storage/ldap/LDAPStorageProvider.java#L1013-L1015

Comment (pedroigor) [reply federation/ldap/src/main/java/org/keycloak/storage/ldap/LDAPStorageProvider.java:995]: Agreed.

File: services/src/main/resources/org/keycloak/userprofile/config/keycloak-default-user-profile.json
Comment (pedroigor) [reply services/src/main/resources/org/keycloak/userprofile/config/keycloak-default-user-profile.json:59]: Works for me. But I'm not sure if having separate groups has a better UX though. We can always improve this later and ask UI team for feedback too.

Comment (mposolda) [reply services/src/main/resources/org/keycloak/userprofile/config/keycloak-default-user-profile.json:59]: @ssilvert @jonkoops @edewit What do you think about this?

In summary: This PR introduces default group "user metadata" into `keycloak-default-user-profile.json`. It is aimed to be used for "metadata attributes" related to user-profile. Not for normal attributes like "firstName", "lastname", "birthday" etc. Examples used in relation to LDAP storage provider are attributes like `LDAP_ID` or `LDAP_ENTRY_DN`. See screenshot above.

Does it makes sense to have separate metadata group like this or should we go with the "flat" structure and treat these attributes as normal attributes?

Comment (ssilvert) [reply services/src/main/resources/org/keycloak/userprofile/config/keycloak-default-user-profile.json:59]: I think having these in a group makes sense.  

We may want to talk about how much control the administrator has over the attribute settings.  For instance, what happens if an admin deletes the attribute from User Profile?

Comment (mposolda) [reply services/src/main/resources/org/keycloak/userprofile/config/keycloak-default-user-profile.json:59]: @ssilvert Administrator has possibility to override the settings for the particular auto-generated attribute as when he define it in user-profile, then the defined version has preference over the auto-generated. This is done here https://github.com/keycloak/keycloak/pull/24282/files#diff-bb4ffa08fae23dc22710d174b81fadf5028b58a8742d9fba4035a2c916e92982R1014 .

So for instance if administrator doesn't want `LDAP_ID` to be shown in the attributes, he can define the `LDAP_ID` attribute in user-profile and disable all permissions. 

The assumption is that most of the administrators will be OK with default settings of "metadata attributes" like LDAP_ID, but there is still possibility to override the default by define it in user profile. 

Is it what you meant?

**Comment by mposolda** [comment]
@pedroigor Updated with added `UserProfileDecorator` to the `keycloak-server-spi` and also adding `realm` as an argument (since most methods on user-storage interfaces like `UserRegistrationProvider`, `UserQueryProvider` and others have realm as an argument instead of relying directly on the realm from the session context). See the latest changes since your last review https://github.com/keycloak/keycloak/pull/24282/commits/3377e9c66ec987b7a83d900b571985a7fbca6ddf .

**Review by pedroigor — DISMISSED** [review]

**Comment by mposolda** [comment]
@keycloak/store-maintainers  Can I please ask store team for review of this one? Regarding store, this PR introduces the interface `UserProfileDecorator`. It allows user-storage providers to "decorate" the metadata provided by user profile realm configuration. It is also implemented on `LDAPStorageProvider` in this PR.

Example usages  - it allows to specify that:

- For LDAP users, the attribute `LDAP_ID` should be shown in the admin console and that it should be read-only

- For read-only LDAP provider, all the user attributes are treated as read-only, which will be also reflected in the UI. This can help to solve the long-standing Keycloak issue, that currently when read-only LDAP user authenticates to account console, all the attributes (firstName, lastName etc) are still shown as writable attributes. But actually user see the error message like "Your account is read-only" once he tries to update something. With this approach, we can show attributes as read-only in the UI.

- Make sure that if some field (like for example `department`) is mandatory for users coming from the user-storage provider XY, then it is marked as "required" also in the UI.

- etc

See screenshots above https://github.com/keycloak/keycloak/pull/24282#issuecomment-1779685210 for how it can look in the admin console.

**Comment by mposolda** [comment]
On hold for now as store team prefer to look at this and discuss in more details.

**Review by vramik — APPROVED** [review]
From store point of view LGTM, thanks @mposolda!

We've discussed possibility to have different namespaces for different types of attributes (e.g. by using prefixes) and it was stated that this change would be non-trivial.

We've also discussed that there are some internal attributes used by Keycloak ([read-only attributes](https://www.keycloak.org/docs/latest/server_admin/index.html#_read_only_user_attributes)). It shouldn't be possible to override them via User Profile: https://github.com/keycloak/keycloak/issues/25047

**Comment by mposolda** [comment]
@vramik Thanks!

**Pull request merged**
